### DISCUSSION TITLE
How do I use the Plugin and settings (beta)

### DISCUSSION THREAD

**Discussion by Vyeche** [vrtmrz/obsidian-livesync#138] [Q&A]
Hi, 

I love the Plugin and setting sync! Great feature. 

Not sure how to use it. I read through the document, but something is missing where I do not see the other device.
- https://github.com/vrtmrz/obsidian-livesync/blob/main/docs/settings.md#plugins-and-settings-beta

I'm curious about how to use the plugin and settings sync. Not sure why replication failed. It doesn't give any more details. I do see some settings for hidden files. Not sure if those need to be enabled while using the beta plugin sync.

Nothing appears on either side of my different device. I have a name given to each device. "Chromebook" and "desktop."

----remote config----
Requesting information to the remote CouchDB has failed. If you are using
IBM Cloudant is the normal behavior.

I noticed that the vault name is not being saved into the data.json

![image](https://user-images.githubusercontent.com/972377/198788327-295a69a4-ff59-49d6-a312-e2293b5ea3e2.png)
![image](https://user-images.githubusercontent.com/972377/198787362-05b2b1fe-3ced-4d71-b38f-0a0a82b10d67.png)
![image](https://user-images.githubusercontent.com/972377/198796961-31ff8a47-42e9-4dbf-9aba-3ee06bbccbc6.png)

![PXL_20221029_123857297.jpg](https://user-images.githubusercontent.com/972377/198832015-dc4e29fd-0ee3-45f2-8d97-2e7a2887f08d.jpg)

**Comment by Vyeche** [comment]
Output from make report.
```json
----remote config----
Requesting information to the remote CouchDB has been failed. If you are using
IBM Cloudant, it is the normal behaviour.

---- Plug-in config ---
couchDB_URI: cloudant
couchDB_USER: 𝑅𝐸𝐷𝐴𝐶𝑇𝐸𝐷
couchDB_PASSWORD: 𝑅𝐸𝐷𝐴𝐶𝑇𝐸𝐷
couchDB_DBNAME: 𝑅𝐸𝐷𝐴𝐶𝑇𝐸𝐷
liveSync: false
syncOnSave: false
syncOnStart: true
savingDelay: 200
lessInformationInLog: true
gcDelay: 0
versionUpFlash: ""
minimumChunkSize: 20
longLineThreshold: 250
showVerboseLog: true
suspendFileWatching: false
trashInsteadDelete: true
periodicReplication: true
periodicReplicationInterval: 60
syncOnFileOpen: true
encrypt: false
passphrase: 𝑅𝐸𝐷𝐴𝐶𝑇𝐸𝐷
workingEncrypt: false
workingPassphrase: 𝑅𝐸𝐷𝐴𝐶𝑇𝐸𝐷
doNotDeleteFolder: false
resolveConflictsByNewerFile: false
batchSave: true
deviceAndVaultName: ""
usePluginSettings: false
showOwnPlugins: false
showStatusOnEditor: true
usePluginSync: false
autoSweepPlugins: false
autoSweepPluginsPeriodic: false
notifyPluginOrSettingUpdated: false
checkIntegrityOnSave: false
batch_size: 250
batches_limit: 40
useHistory: true
disableRequestURI: true
skipOlderFilesOnSync: true
checkConflictOnlyOnOpen: true
syncInternalFiles: false
syncInternalFilesBeforeReplication: false
syncInternalFilesIgnorePatterns: \/node_modules\/, \/\.git\/, \/obsidian-livesync\/
syncInternalFilesInterval: 60
additionalSuffixOfDatabaseName: ""
ignoreVersionCheck: false
lastReadUpdates: 16
deleteMetadataOfDeletedFiles: false
syncIgnoreRegEx: ""
syncOnlyRegEx: ""
customChunkSize: 0
readChunksOnline: true
watchInternalFileChanges: true
automaticallyDeleteMetadataOfDeletedFiles: 0

```

**Reply by vrtmrz** [reply]
The `Device and vault name` is not saved into the `data.json` now.
But, this feature is now realised by the `Sync hidden files`. (Getting be obsoleted but kept for compatibility).
If we want to keep old versions of plugins, [Screwdriver](https://github.com/vrtmrz/obsidian-screwdriver) will help us.
Could you please try `Sync hidden files` once?

**Reply by Vyeche** [reply]
Do you mean that a new vault should not use plugin sync (beta)? All I had to do was enable `Sync hidden files`?

**Reply by vrtmrz** [reply]
Depends on your requirements.
If you want to control the versions of plugins on each device, `Plugins and their settings` should be useful.
Or just only to synchronise all of them, `Sync hidden files` would be the straight way.

Sync hidden files synchronise all plugins and detect the changes, and ask you to reload or ignore them.

Note: In the `Sync hidden files` way, we can not refuse the change, but I found we usually do not. Instead, we can revert the changes from `Pick a file to show history`.

**Reply by Vyeche** [reply]
So the best way is to use `Sync hidden files`? The other feature is kept for compatibility and might not be developed. Is that correct?

**Reply by Vyeche** [reply]
Hmm, it syncs all my git. Not good. I thought the regex covered that.

![image](https://user-images.githubusercontent.com/972377/198916574-8ff7105a-38fc-406c-8a46-eb223ab3d230.png)

**Reply by Vyeche** [reply]
Does a skip pattern mean include git? If I want to exclude git, I need to remove it from that box? Right? @vrtmrz

**Reply by vrtmrz** [reply]
Sorry for being late @Vyeche! The default configuration has not been enough.

In the default, `\/node_modules\/, \/\.git\/, \/obsidian-livesync\/`. but it must be `\/node_modules\/, (^|\/)\.git\/, \/obsidian-livesync\/` to contain the `.git` folder on top of our vault.

**Reply by Vyeche** [reply]
@vrtmrz I think that did the trick. However, I see an odd error from before about plugin sync. Any idea why replication failed? 
- It starts to show that error when I enable hidden files.
- **Then, the sync across all devices breaks for me. Forcing me to rebuild again.**

![image](https://user-images.githubusercontent.com/972377/199137445-fe179c99-b650-4d7e-bc42-014190cd3035.png)

**Reply by Vyeche** [reply]
@vrtmrz Appears to be coming from LocalPouchDB.ts 
- I saw the same error earlier with using the beta plugin. 
- https://github.com/vrtmrz/obsidian-livesync/discussions/138#discussion-4516595
https://github.com/vrtmrz/obsidian-livesync/blob/88311659655122e68b66a3856089398eeeb4745e/src/LocalPouchDB.ts
![image](https://user-images.githubusercontent.com/972377/199139516-b1df9036-fc6c-4274-8080-fa999f61bd48.png)

**Reply by vrtmrz** [reply]
It is indicated when the request payload of the request becomes larger than the size which Cloudant accepts. 

In that case, the batch size and batch limit will be reduced and the request will be retried while in replication. Otherwise, the plug-in will say he has given it up.

Would you please set a lower value `Batch limit` or `Batch size` once and try replication again?
I missed the making logs that what request makes the error.

**Reply by Vyeche** [reply]
Hmm it failed at the default values.

What should I set it? @vrtmrz 

![Screenshot_20221031-221253.png](https://user-images.githubusercontent.com/972377/199143693-ab8ffb4f-3724-4455-b7b5-562bfce0af40.png)

**Reply by Vyeche** [reply]
Perhaps the logs would give us more insight?

**Reply by Vyeche** [reply]
@vrtmrz I set it to 60 Batch size, which seems to have corrected the issue. 

However, I do see a workflow issue. It's consistently asking me to reload my plugins depending on which devices I have open simultaneously. Meanwhile, I set the scan hidden files on my Chromebook and desktop computer.

**Reply by vrtmrz** [reply]
Sorry to have left the keyboard so abruptly.  And thank you for checking with several configuration values. If the vault could be replicated once, we can restore the value to the default.

Note: A schematic table is shown below.

| Seq  | Size of chunk |  BL 3   | BL 2  |
| :--- | :------------ | :-----: | :---: |
| 1    | 10            |   90    |  30   |
| 2    | 20            |   \|    |   ↥   |
| 3    | 60            |    ↥    |  90   |
| 4    | 30            | **120** |   ↥   |
| 5    | 40            | **\|**  |  80   |
| 6    | 40            |  **↥**  |   ↥   |

If the Batch limit is set to 3 (which means 250 in the original configuration), seq 1-3 and seq 4-6 are sent in one batch each and reach 90 and 120 respectively.
This would cause the second request to fail if the payload size is limited to 100 (the reason why it could not be detected needs to be verified).

So, by setting the batch limit to 2 (You know, 60.) and keeping the chunk size to 90, the limit is circumvented. (However, I had intended to design it so that it would be done automatically.)

> However, I do see a workflow issue. It's consistently asking me to reload my plugins depending on which devices I have open simultaneously. Meanwhile, I set the scan hidden files on my Chromebook and desktop computer.

I think it is happening with certain plugins. May I ask what plugin said that desired to be reloaded?

**Reply by Vyeche** [reply]
@vrtmrz 

The data.json file gets changed for plugins. I understand what you mean. So that's why the message comes. Right? 

The issue I noticed is that the plugins turn off. 

![Screenshot_20221102-111952.png](https://user-images.githubusercontent.com/972377/199529192-d7ab929f-c41e-4f6e-a25c-8b093b359bd5.png)

**Reply by Vyeche** [reply]
I need to spend some more time understanding the terms. In the mean time I got an idea.

Is it possible for the chunk and batch size to be adaptive based on your table? 

Say for example when it fails it adjusts a few times to correct itself until it files success and then sets up to the correct working settings. That way the user doesn't need to fiddle with that area.

Perhaps you can add a checkbox for adaptive chunking and batching. 

What do you think? @vrtmrz

**Reply by vrtmrz** [reply]
Thank you for the screenshot!
Maybe a feature like `Reload specific plugins automatically`will be useful, if these plugins can be reloaded immediately and no side effects (LiveSync and TagFolder are slow-stater so they will be treated annually )

> Is it possible for the chunk and batch size to be adaptive based on your table?

Completely as you said is right!
Already LiveSync (might) detects that payload size is exceeded, and (in design) it reduces batch limit to adopt automatically. But I split them, so we have to investigate more.

**Reply by Vyeche** [reply]
I noticed that even after I adjust the numbers, it will continue to fail seems tricky to figure out on my own. So we need some algorithms to keep trying. @vrtmrz

**Reply by vrtmrz** [reply]
Thank you so much for your cooperation May I see a screenshot of the network pane?
I want to know what URI makes this issue.

**Comment by Vyeche** [comment]
After removing the plugin something odd happened.
document_too_large:i:.obsidian/plugins/obsidian-admonition/main.js
Installing the app gets rid of the error.
Adjusting the batch size removed the error afterword. 

![image](https://user-images.githubusercontent.com/972377/199535072-0fb19855-3ce1-471e-81c8-a2f01061bfb5.png)

**Reply by vrtmrz** [reply]
This might be the same as https://github.com/vrtmrz/obsidian-livesync/discussions/138#discussioncomment-4034669 .

**Reply by Vyeche** [reply]
It was breaking so much in the morning today. No matter how I set the limit on the batch, it had syncing problems and brought in plugins that I removed. I will get more screenshots of the network later. @vrtmrz 
![image](https://user-images.githubusercontent.com/972377/199864525-123f6011-6a0c-427d-95df-236b758de499.png)
![image](https://user-images.githubusercontent.com/972377/199864609-31af2435-5b78-4ee6-b362-989df04ccdd8.png)
![image](https://user-images.githubusercontent.com/972377/199864628-ee4e3dc0-d368-47a5-b726-0487c827a78a.png)

**Reply by Vyeche** [reply]
@vrtmrz more blocked requests
![image](https://user-images.githubusercontent.com/972377/199995276-80d7bf08-3d6c-4c76-944d-4ccedca761f6.png)
![image](https://user-images.githubusercontent.com/972377/199995384-edc35a38-5951-4805-bbb3-690f3cd06aa8.png)

**Reply by vrtmrz** [reply]
Thank you for your reply (Image saved. And please delete it, it contains the URL).
I have been investigating this for a couple of hours now, but haven't reproduced it yet.
If you see the red-lined error again (as below), could you please check the `Response` pane?
I have seen when amount of my requests reached the rate limit.

![image](https://user-images.githubusercontent.com/45774780/200013149-b1409e43-70e4-4f7c-ace1-6bf2085766e5.png)

**Reply by Vyeche** [reply]
Okay, I deleted it; I will keep looking for more in the response pane.

**Reply by Vyeche** [reply]
The notification for the plugin gets too much. 
Perhaps a way to toggle it off.

![Screenshot_20221104-165137.png](https://user-images.githubusercontent.com/972377/200072696-3994c4d5-fca0-4d8a-8a83-6ca3f321a4d6.png)

**Reply by Vyeche** [reply]
I see some 413 errors, but nothing interesting on network @vrtmrz 
I noticed 'replication failure' when I had multiple devices syncing.

What's worse is that it broke the live sync (not plugin sync if they are different), or at least I noticed that my files look older than my other device files.
![image](https://user-images.githubusercontent.com/972377/200081581-300d5d3e-00e6-4814-a4e2-b313bde3d44c.png)

**Reply by Vyeche** [reply]
@vrtmrz I thought, what if the plugins had a single point of truth, let's say, my home desktop? Every time I update that area, my other devices will get the update instead of any device changing. I don't see myself adjusting my plugins as much as much notes. Perhaps that will remove these sync issues?

**Reply by vrtmrz** [reply]
> Disabling the notification

It will be implemented soon! (Sorry, I will usually be a good dad on holidays). In that case, Just ignoring or automatically reloading should be configured.
However, perhaps we can avoid them by skipping `data.json` in synchronisation.

But sorry for breaking synchronisations. There must be some deeper problems. I'd love to thank you once again.
If file contents are the same, they will not be overwritten, so perhaps their timestamps look miss matched.

The 413 in the screenshot is apparently not in the list above. Which request was it?
This is a totally different solution, but what if I disable `Read chunks online`?

> I thought, what if the plugins had a single point of truth, 

Oh! This is the functionality that hidden file sync lacks and Plugins and their settings have. 

I thought that almost everything can be solved by Sync hidden file. But these use-cases might be useful.
Still need some fixes, but it will not be `obsoleted` now.  Thank you for your patience and for pointing me out!

**Reply by Vyeche** [reply]
@vrtmrz 

That's great; thank you for the reply! 

I will pause using the plugin sync; it is hard to use my vault productively. I will wait for some new updates to test later. I am looking forward to new updates.

Sorry for bugging you.

**Comment by vrtmrz** [comment]
Before fixing the plugin and settings, I fixed the feature which adjusts the `batch limit` automatically.
(Maybe it also fixes the `plugin and settings`.) 

If some of the chunks are still big, we have to rebuild the database again. But it is also be detected.

It would be great if you could update it and give it a try.

**Reply by Vyeche** [reply]
Okay, I'll plan something maybe during the week or next because it was hard to stay productive while troubleshooting the plugins.

@vrtmrz 
Is it possible for you can create another IBM test account, so I can quickly make a new vault and try it during the week?

**Reply by vrtmrz** [reply]
Thank you so much! I also think so, I really appreciate your significant contributions every time!

> Is it possible for you can create another IBM test account, so I can quickly make a new vault and try it during the week?

I'm really sorry for telling you this, but I also could not make another Cloudant account. We can use different database names in the same Cloudant instance to split the influence. Could this help us?

Or fly.io could be very easy to set up and initialise. 
Now it doesn't require installing clients locally. (I have found out the way to setting CouchDB up from google Colab)

**Reply by GSanrio** [reply]
In the new version of the plugin, there is no longer an item for selecting plugin synchronization in the custom settings. How should I set it up?
